rhino/src/main/java/org/mozilla/javascript/nat/type/TypeInfoBase.java

```diff
@@ -0,0 +1,37 @@
+package org.mozilla.javascript.nat.type;
+
+import java.lang.reflect.Array;
+
+public abstract class TypeInfoBase implements TypeInfo {
+    private TypeInfo asArray;
+    private Object emptyArray;
```

Comment:
these classes should be immutable as they are in the cache. So every field should be final or initialization should be synchronized.